fix: enable rendering of commands properties inside CommandsWrapper

mgabelle opened the pull request:
enable the rendering of Property<List<String>> commands, interpreter, and before commands within the commands wrapper to have access to workingDir variable

loicmathieu commented on core/src/main/java/io/kestra/core/models/property/Property.java:
```diff
@@ -61,6 +61,18 @@ String getExpression() {
+     * A property build with this method will need to be rendered.
+     * This method can be used when for testing and to set default value like pebble template that will require rendering.
+     */
+    public static <V> Property<V> ofValueNotRendered(V value) {
```
Please remove this and use `new Property(expression)`

mgabelle replied:
Removed

Commit:
fix: move commands to Property

migrate to Property in TaskCommands and CommandsWrapper
implement beforeCommand and interpreter

mgabelle commented:
Hi @loicmathieu ,
If you have time of coure would you be able to review/validate and eventually merge this PR tomorrow ?
This way I would be able to use it on Monday to prepare the bugfix release of the script plugins

loicmathieu commented on core/src/main/java/io/kestra/plugin/core/runner/Process.java:
```diff
@@ -133,11 +136,14 @@ public TaskRunnerResult<TaskRunnerDetailResult> run(RunContext runContext, TaskC
         processBuilder.directory(taskCommands.getWorkingDirectory().toFile());
-        processBuilder.command(taskCommands.getCommands());
+
+        List<String> renderedCommands = this.renderCommands(runContext, taskCommands);
```
Reading the code in CommandsWrapper, it looks like commands are already rendered inside the `command` attribute. Either use this or remove it.

mgabelle replied:
Replaced with `runContext.render(taskCommands.getCommands()).asList(String.class)`

loicmathieu commented on core/src/main/java/io/kestra/core/models/tasks/runners/TaskRunner.java:
```diff
@@ -130,6 +131,24 @@ public String toAbsolutePath(RunContext runContext, TaskCommands taskCommands, S
+            .flatMap(Collection::stream).toList();
+    }
+
+    public List<String> renderCommandsFromList(RunContext runContext, TaskCommands taskCommands, Property<List<String>> commands) throws IllegalVariableEvaluationException, IOException {
```
```suggestion
    private List<String> renderCommandsFromList(RunContext runContext, TaskCommands taskCommands, Property<List<String>> commands) throws IllegalVariableEvaluationException, IOException {
```

It seems it's only used by the `renderCommands` method so better to make it private.

mgabelle replied:
Made method `private`

Commit:
fix: PR comments

Bdlily commented:
## PR Code Suggestions ✨

<!-- e6819f3 -->

<table><thead><tr><td><strong>Category</strong></td><td align=left><strong>Suggestion&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></td><td align=center><strong>Impact</strong></td></tr><tbody><tr><td rowspan=1>Possible issue</td>
<td>



<details><summary>Add null filter to prevent NPE</summary>

___


**The <code>renderCommands</code> method may produce incorrect command ordering when any of the <br>property lists are null. When <code>taskCommands.getCommands()</code>, <code>getBeforeCommands()</code>, or <br><code>getInterpreter()</code> return null, <code>renderCommandsFromList</code> returns null (since <code>commands == </code><br><code>null</code> check returns null, not empty list), causing <code>flatMap(Collection::stream)</code> to <br>throw NullPointerException. Ensure null safety by filtering null lists before <br>flattening.**

[core/src/main/java/io/kestra/core/models/tasks/runners/TaskRunner.java [134-141]](https://github.com/kestra-io/kestra/pull/7381/files#diff-4d8d537fedbddb0d7aa16677b5abc50de4101cba1ddaf9f2f15492524f72850bR134-R141)

```diff
 public List<String> renderCommands(RunContext runContext, TaskCommands taskCommands) throws IllegalVariableEvaluationException, IOException {
     List<String> renderedCommands = this.renderCommandsFromList(runContext, taskCommands, taskCommands.getCommands());
     List<String> renderedBeforeCommands = this.renderCommandsFromList(runContext, taskCommands, taskCommands.getBeforeCommands());
     List<String> renderedInterpreter = this.renderCommandsFromList(runContext, taskCommands, taskCommands.getInterpreter());
 
     return Stream.of(renderedInterpreter, renderedBeforeCommands, renderedCommands)
+        .filter(Objects::nonNull)
         .flatMap(Collection::stream).toList();
 }
```
<details><summary>Suggestion importance[1-10]: 8</summary>

__

Why: The suggestion correctly identifies a potential `NullPointerException` when `renderCommandsFromList` returns null (which happens when `commands == null` in `ScriptService.replaceInternalStorage`). Adding `.filter(Objects::nonNull)` before `.flatMap(Collection::stream)` is a valid fix for this edge case. However, the score is not higher because the PR code already handles null in `ScriptService.replaceInternalStorage` by returning `Collections.emptyList()`, so the NPE would only occur if the `TaskCommands` getters return null `Property` objects, which is less likely in practice.


</details></details></td><td align=center>Medium

</td></tr><tr><td rowspan=1>General</td>
<td>



<details><summary>Fix parameter shadowing and consistency</summary>

___


**The <code>renderCommands</code> method parameter shadows the class field <code>commands</code>, which is <br>confusing and error-prone. Additionally, this method uses <code>this.runContext</code> but <br>receives <code>runContext</code> as parameter - it should use the passed parameter for <br>consistency. Rename the parameter to avoid shadowing and use the passed <code>runContext</code> <br>parameter.**

[script/src/main/java/io/kestra/plugin/scripts/exec/scripts/runners/CommandsWrapper.java [268-274]](https://github.com/kestra-io/kestra/pull/7381/files#diff-9307a25d3a118c6f6a5c80f13555830678b0dcae5463e0e35987e43277653ff9R268-R274)

```diff
-public List<String> renderCommands(RunContext runContext, Property<List<String>> commands) throws IllegalVariableEvaluationException, IOException {
+public List<String> renderCommands(RunContext runContext, Property<List<String>> commandsProperty) throws IllegalVariableEvaluationException, IOException {
     TaskRunner<?> taskRunner = this.getTaskRunner();
     return ScriptService.replaceInternalStorage(
-        this.runContext,
+        runContext,
         taskRunner.additionalVars(runContext, this),
-        commands,
+        commandsProperty,
         taskRunner instanceof RemoteRunnerInterface
     );
 }
```
<details><summary>Suggestion importance[1-10]: 6</summary>

__

Why: The suggestion correctly identifies two issues: the parameter `commands` shadows the class field `commands`, and the method uses `this.runContext` instead of the passed `runContext` parameter. Both are valid code quality concerns that could lead to bugs. However, this is a minor refactoring suggestion rather than a critical bug fix, as the current code likely works correctly in practice.

</details></details></td><td align=center>Low

</td></tr></tr></tbody></table>
